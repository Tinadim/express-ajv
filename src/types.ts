import { RequestHandler, Request } from 'express';
import Ajv, { Options as AjvOptions } from 'ajv';

export interface RequestSchema {
    type: 'object';
    properties: {
        body?: object;
        params?: object;
        query?: object;
    };
    required?: string[];
    [key: string]: any;
}

export interface ValidatorOptions {
    ajv?: Ajv.Ajv;
    ajvOptions?: AjvOptions;
    formatError?: (error: any) => any;
    getContextParams?: (req: Request) => any;
}

export type SchemaCollection = { [schemaName: string]: RequestSchema };

export interface ValidatorMiddleware {
    ajv: Ajv.Ajv;
    (schemaName: string): RequestHandler;
}
