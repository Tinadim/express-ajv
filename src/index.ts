export {
    SchemaCollection,
    ValidatorMiddleware,
    ValidatorOptions,
} from './types';

export { createValidator } from './create-validator';
