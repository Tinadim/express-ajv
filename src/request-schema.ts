import { RequestSchema } from './types';

function createSchema(querySchema?: object, bodySchema?: object, paramsSchema?: object) {
    const schema: { [key: string]: any } = {
        type: 'object',
        properties: {},
        required: [],
    };

    if (querySchema) {
        schema.required.push('query');
        schema.properties['query'] = querySchema;
    }

    if (bodySchema) {
        schema.required.push('body');
        schema.properties['body'] = bodySchema;
    }

    if (paramsSchema) {
        schema.required.push('params');
        schema.properties['params'] = paramsSchema;
    }

    return schema as RequestSchema;
}

export function createGetRequestSchema(querySchema: object, paramsSchema?: object) {
    return createSchema(querySchema, undefined, paramsSchema);
}

export function createPostRequestSchema(bodySchema: object, paramsSchema?: object) {
    return createSchema(undefined, bodySchema, paramsSchema);
}
