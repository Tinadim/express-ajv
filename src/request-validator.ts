import { forOwn, pick } from 'lodash';
import { RequestHandler, Request, Response, NextFunction } from 'express';
import Ajv, { ValidateFunction } from 'ajv';
import { SchemaCollection, ValidatorOptions } from './types';

export class RequestValidator {
    ajv: Ajv.Ajv;
    private validators: { [schemaName: string]: ValidateFunction } = {};
    private contextExtractor: (req: Request) => any;

    constructor(
        schemas: SchemaCollection,
        options?: ValidatorOptions,
    ) {
        this.ajv = options?.ajv ?? new Ajv(options?.ajvOptions);
        this.contextExtractor = options?.contextExtractor ?? this.defaultValidationContext;
        this.registerSchemas(schemas);
    }

    private registerSchemas(schemaCollection: { [schemaName: string]: object }) {
        forOwn(schemaCollection, (schema: object, schemaName: string) => {
            const validator = this.ajv.compile(schema);
            this.validators[schemaName] = validator;
        });
    }

    private getValidatorForSchema(schemaName: string) {
        if (!schemaName) {
            throw new Error('Please provide the schema name');
        }

        const validator = this.validators[schemaName];
        if (!validator) {
            throw new Error(`Validator for schema with name ${schemaName} not found`);
        }
        return validator;
    }

    private defaultValidationContext(req: Request) {
        return {};
    }

    validateSchema(schemaName: string) {
        const validator = this.getValidatorForSchema(schemaName);

        const middleware: RequestHandler = (req: Request, res: Response, next: NextFunction) => {
            const contextParams = this.contextExtractor(req);
            const context = Object.assign({}, this.ajv, contextParams);
            const dataToValidate = pick(req, ['body', 'params', 'query']);

            return this.validate(dataToValidate, validator, context)
                .then(() => next())
                .catch((error) => {
                    next(error);
                });
        };

        return middleware;
    }

    private validate(data: any, validator: ValidateFunction, context: any) {
        return validator.$async
            ? this.validateAsync(data, validator, context)
            : this.validateSync(data, validator, context);
    }

    private validateSync(data: any, validator: ValidateFunction, context: any) {
        const valid = validator.call(context, data);
        return valid
            ? Promise.resolve()
            : Promise.reject(new Ajv.ValidationError(validator.errors!));
    }

    private validateAsync(data: any, validator: ValidateFunction, context: any) {
        return validator.call(context, data) as Promise<any>;
    }
}
