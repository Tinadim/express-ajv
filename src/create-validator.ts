import { isObject } from 'lodash';
import { SchemaCollection, ValidatorMiddleware, ValidatorOptions } from './types';
import { RequestValidator } from './request-validator';

export function createValidator(schemas: SchemaCollection, options?: ValidatorOptions): ValidatorMiddleware {
    if (!isObject(schemas)) {
        throw new Error('schemas should be an object');
    }

    const validator = new RequestValidator(schemas, options);
    const validatorMiddleware = (schema: string) => validator.validateSchema(schema);
    return Object.assign(validatorMiddleware, { ajv: validator.ajv }) as ValidatorMiddleware;
}
