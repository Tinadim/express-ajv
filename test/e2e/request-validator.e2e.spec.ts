import http from 'http';
import bodyParser from 'body-parser';
import express, { Application, NextFunction, Request, Response } from 'express';
import { merge, pick } from 'lodash';
import supertest from 'supertest';
import { createValidator, ValidatorOptions } from '../../src';
import { schemas } from '../fixtures/schemas';

describe('request-validator.e2e-spec.ts', () => {
    let app: Application;
    let server: http.Server;

    const port = 8080;
    const defaultOptions: ValidatorOptions = {
        ajvOptions: {
            allErrors: true,
            format: 'full',
        },
    };
    const badData = {
        name: 'Be',
        age: -1,
    };
    const validData = {
        name: 'Bernardo',
        age: 94,
        city: 'new-york',
    };

    afterEach(() => {
        return new Promise((resolve) => {
            server.close(resolve);
        });
    });

    function errorFormatter(error?: any) {
        const details = error.errors.map((validationError: any) => ({
            field: validationError.dataPath,
            message: validationError.message,
        }));
        const formattedError = {
            details,
            message: `Original error message: ${error.message}`,
        };
        return formattedError;
    }

    function errorHandler(err: any, req: Request, res: Response, next: NextFunction) {
        const error = pick(err, ['message', 'errors', 'details']);
        res.status(400).send({ error });
    }

    function createExpressApp(options?: ValidatorOptions) {
        app = express();
        const mergedOptions = merge({}, defaultOptions, options);
        const validate = createValidator(schemas, mergedOptions);

        app.use(bodyParser.urlencoded({ extended: true }));
        app.use(bodyParser.json());

        app.get('/test/:userId', validate('getRequest'), (req: Request, res: Response) => {
            res.status(200).send({ success: true });
        });

        app.post('/test/:userId', validate('postRequest'), (req: Request, res: Response) => {
            res.status(200).send({ success: true });
        });

        app.use('/', errorHandler);

        return new Promise((resolve) => {
            server = app.listen(port, resolve);
        });
    }

    async function testRequest(method: 'get' | 'post', url: string,  data: object) {
        const request = supertest(app)[method](url);
        if (method === 'get') {
            request.query(data);
        } else {
            request.send(data);
        }

        const response = await request;
        expect(response.body).toMatchSnapshot();

        return response;
    }

    async function testInvalidRequest(method: 'get' | 'post') {
        const url = '/test/1234';
        const response = await testRequest(method, url, badData);
        expect(response.status).toBe(400);
    }

    async function testValidRequest(method: 'get' | 'post') {
        const url = '/test/576871e16b01de9500a46269';
        const response = await testRequest(method, url, validData);
        expect(response.status).toBe(200);
    }

    describe('GET requests', () => {
        it('Should return an error if the request does not comply with the provided schema', async () => {
            await createExpressApp();
            await testInvalidRequest('get');
        });

        it('Should accept valid requests', async () => {
            await createExpressApp({ ajvOptions: { coerceTypes: true } });
            await testValidRequest('get');
        });
    });

    describe('POST requests', () => {
        it('Should return an error if the request does not comply with the provided schema', async () => {
            await createExpressApp();
            await testInvalidRequest('post');
        });

        it('Should accept valid requests', async () => {
            await createExpressApp();
            await testValidRequest('post');
        });
    });
});
