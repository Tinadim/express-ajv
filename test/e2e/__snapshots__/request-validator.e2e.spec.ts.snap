// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`request-validator.e2e-spec.ts GET requests Should accept valid requests 1`] = `
Object {
  "success": true,
}
`;

exports[`request-validator.e2e-spec.ts GET requests Should return an error if the request does not comply with the provided schema 1`] = `
Object {
  "error": Object {
    "errors": Array [
      Object {
        "dataPath": ".params.userId",
        "keyword": "pattern",
        "message": "should match pattern \\"[a-f0-9]{24}\\"",
        "params": Object {
          "pattern": "[a-f0-9]{24}",
        },
        "schemaPath": "#/properties/params/properties/userId/pattern",
      },
      Object {
        "dataPath": ".query.name",
        "keyword": "minLength",
        "message": "should NOT be shorter than 3 characters",
        "params": Object {
          "limit": 3,
        },
        "schemaPath": "#/properties/query/properties/name/minLength",
      },
      Object {
        "dataPath": ".query.age",
        "keyword": "type",
        "message": "should be number",
        "params": Object {
          "type": "number",
        },
        "schemaPath": "#/properties/query/properties/age/type",
      },
      Object {
        "dataPath": ".query",
        "keyword": "required",
        "message": "should have required property 'city'",
        "params": Object {
          "missingProperty": "city",
        },
        "schemaPath": "#/properties/query/required",
      },
    ],
    "message": "validation failed",
  },
}
`;

exports[`request-validator.e2e-spec.ts GET requests Should return the error formatted in the appropriate format 1`] = `
Object {
  "error": Object {
    "details": Array [
      Object {
        "field": ".params.userId",
        "message": "should match pattern \\"[a-f0-9]{24}\\"",
      },
      Object {
        "field": ".query.name",
        "message": "should NOT be shorter than 3 characters",
      },
      Object {
        "field": ".query.age",
        "message": "should be number",
      },
      Object {
        "field": ".query",
        "message": "should have required property 'city'",
      },
    ],
    "message": "Original error message: validation failed",
  },
}
`;

exports[`request-validator.e2e-spec.ts POST requests Should accept valid requests 1`] = `
Object {
  "success": true,
}
`;

exports[`request-validator.e2e-spec.ts POST requests Should return an error if the request does not comply with the provided schema 1`] = `
Object {
  "error": Object {
    "errors": Array [
      Object {
        "dataPath": ".params.userId",
        "keyword": "pattern",
        "message": "should match pattern \\"[a-f0-9]{24}\\"",
        "params": Object {
          "pattern": "[a-f0-9]{24}",
        },
        "schemaPath": "#/properties/params/properties/userId/pattern",
      },
      Object {
        "dataPath": ".body.name",
        "keyword": "minLength",
        "message": "should NOT be shorter than 3 characters",
        "params": Object {
          "limit": 3,
        },
        "schemaPath": "#/properties/body/properties/name/minLength",
      },
      Object {
        "dataPath": ".body.age",
        "keyword": "minimum",
        "message": "should be >= 0",
        "params": Object {
          "comparison": ">=",
          "exclusive": false,
          "limit": 0,
        },
        "schemaPath": "#/properties/body/properties/age/minimum",
      },
      Object {
        "dataPath": ".body",
        "keyword": "required",
        "message": "should have required property 'city'",
        "params": Object {
          "missingProperty": "city",
        },
        "schemaPath": "#/properties/body/required",
      },
    ],
    "message": "validation failed",
  },
}
`;

exports[`request-validator.e2e-spec.ts POST requests Should return the error formatted in the appropriate format 1`] = `
Object {
  "error": Object {
    "details": Array [
      Object {
        "field": ".params.userId",
        "message": "should match pattern \\"[a-f0-9]{24}\\"",
      },
      Object {
        "field": ".body.name",
        "message": "should NOT be shorter than 3 characters",
      },
      Object {
        "field": ".body.age",
        "message": "should be >= 0",
      },
      Object {
        "field": ".body",
        "message": "should have required property 'city'",
      },
    ],
    "message": "Original error message: validation failed",
  },
}
`;
